include/deal.II/base/mpi_compute_index_owner_internal.h

```diff
@@ -35,6 +35,39 @@ namespace Utilities
        */
       namespace ComputeIndexOwner
       {
+        class FlexibleIndexStorage
+        {
+        public:
+          using index_type = unsigned int;
+          static const index_type invalid_index_value =
+            numbers::invalid_unsigned_int;
+
+          FlexibleIndexStorage(const bool use_vector = true);
+
+          void
+          reinit(const bool use_vector, const std::size_t size);
+
+          void
+          fill(const std::size_t start,
+               const std::size_t end,
+               const index_type &value);
+
+          index_type &
+          operator[](const std::size_t index);
+
+          index_type
+          operator[](const std::size_t index) const;
+
+          bool
+          entry_has_been_set(const std::size_t index) const;
+
+        private:
+          bool                              use_vector;
+          std::size_t                       size;
+          std::vector<index_type>           data;
+          std::map<std::size_t, index_type> data_map;
+        };
```

Comment:
Now also available as `std_cxx17::variant` (see #13497).

Reply:
> I think you can make this easier if you used std::variant/boost::variant instead of reinventing the corresponding wheel in FlexibleIndexStorage.

@bangerth Thanks for the PR. As far a I see, `std::variant` would only remove a single bool and replace the if-statements by something more verbose. Since, the `std::map` is a quick solution for the reduction of the memory consumption and might change in the future, I would suggest to keep it like it is.

Reply:
I like things such as `std::variant` because they use standardized language every programmer will understand. I don't have to read through the details of the class to understand how it works (it's relatively sparse on documentation right now :-) ). The value of standard idioms isn't always that it makes code shorter, but sometimes also that it makes it easier to *talk about* because it uses standardized language.

I'll leave it up to you what you want to do with the suggestion, though :-)

Reply:
Personally, I would prefer to stick to the bool.